crates/nu-cli/src/types/deduction.rs

```diff
@@ -0,0 +1,1451 @@
+                        if let Some(positional) = &internal.args.positional {
+                            //Infer shapes in positional
+                            self.infer_shapes_based_on_signature_positional(
+                                positional, &signature,
+                            )?;
+                        }
+                        if let Some(named) = &internal.args.named {
+                            //Infer shapes in named
+                            self.infer_shapes_based_on_signature_named(named, &signature)?;
+                        }
+                    }
+                    //vars in expressions can be deduced by their usage
+                    //e.G. 1..$var ($var is of type Int)
+                    if let Some(positional) = &internal.args.positional {
+                        //Infer shapes in positional
+                        for (_pos_idx, pos_expr) in positional.iter().enumerate() {
+                            self.infer_shapes_in_expr(
+                                (cmd_pipeline_idx, pipeline),
+                                pos_expr,
+                                registry,
+                            )?;
+                        }
+                    }
+                    if let Some(named) = &internal.args.named {
+                        trace!("Infering vars in named exprs");
+                        for (_name, val) in named.iter() {
+                            if let NamedValue::Value(_, named_expr) = val {
+                                self.infer_shapes_in_expr(
+                                    (cmd_pipeline_idx, pipeline),
+                                    named_expr,
+                                    registry,
+                                )?;
+                            }
+                        }
+                    }
+                }
+                ClassifiedCommand::Expr(spanned_expr) => {
+                    trace!(
+                        "Infering shapes in ClassifiedCommand::Expr: {:?}",
+                        spanned_expr
+                    );
+                    self.infer_shapes_in_expr(
+                        (cmd_pipeline_idx, pipeline),
+                        spanned_expr,
+                        registry,
+                    )?;
+                }
+                ClassifiedCommand::Dynamic(_) | ClassifiedCommand::Error(_) => unimplemented!(),
+            }
+        }
+        Ok(())
+    }
+
+    fn infer_shapes_based_on_signature_positional(
+        &mut self,
+        positionals: &[SpannedExpression],
+        signature: &Signature,
+    ) -> Result<(), ShellError> {
+        trace!("Infering vars in positionals");
+        //TODO currently correct inference for optional positionals is not implemented.
+        // See https://github.com/nushell/nushell/pull/2486 for a discussion about this
+        // For now we assume every variable in an optional positional is used as this optional
+        // argument
```

Comment:
Can you say a bit more about "we assume every variable in an optional positional is used as this optional argument"? I'm not sure what this means.

Reply:
As far as I understand the current parser implementation:
[The positional position is always updated by one](https://github.com/nushell/nushell/blob/38ccbef167bf288f40428e718d9cb6d530816389/crates/nu-parser/src/parse.rs#L1246-L1262), regardless whether the positional fitted the (possible) optional positional [link](https://github.com/nushell/nushell/blob/38ccbef167bf288f40428e718d9cb6d530816389/crates/nu-parser/src/parse.rs#L1151-L1157)

So If there would be a variable in an optional positional place, the parser would assume it to fit this positional and so does the deduction module.

I will update the comment to make it more clear :)